### DISCUSSION TITLE
Communication issue between SX1276 and SX1262

### DISCUSSION THREAD

**Discussion by lucadjc** [jgromes/RadioLib#817] [Q&A] [answered]
Hello,
I've a Heltec v2 Lora device based on SX1276 and that is acting in my scenario as sender.
I've a new Heltec v3 Lora device based  on SX1262 device that is acting as receiver.

lora seems up & running but messages are not received, I don't know what wrongly i'm doing.
thanks in advance for your support

### **Sender**
```c++
#include <ArduinoJson.hpp>
#include <ArduinoJson.h>
#include <Adafruit_Sensor.h>
#include <DHT.h>
#include <ESPDateTime.h>
#include <RadioLib.h>
#include <SPI.h>
#include <Hash.h>


// - Pin configs - ESP32 -
#define ss 18   
#define rst  14 
#define dio0 26 
#define dio1 35 
SX1276 radio = new Module(ss, dio1, rst, dio0);

#define DHTPIN 4     // what digital pin the DHT22 is conected to
#define DHTTYPE DHT22   // there are multiple kinds of DHT sensors

DHT dht(DHTPIN, DHTTYPE);

int timeSinceLastRead = 0;
int tempAvailable = 0;
char temperatureMessage[256];

#define uS_TO_S_FACTOR 1000000  /* Conversion factor for micro seconds to seconds */
#define TIME_TO_SLEEP  600        /* Time ESP32 will go to sleep (in seconds) */
RTC_DATA_ATTR int bootCount = 0;

//Used for read battery data
#define VBAT_PIN 39
#define BATTV_MAX    4.14     // maximum voltage of battery
#define BATTV_MIN    1.3     // what we regard as an empty battery
#define BATTV_LOW    1.06     // voltage considered to be low battery
int battery_level = 0;
int pin_value = 0;

void setup() {
    initSerial();
    initDHT22();
    initLoRaSender();
    batteryStatus();
    manageTemperature();
    //startSleep();
    
}

void batteryStatus() {
    pin_value = analogRead(VBAT_PIN);
    //Serial.println((String)"Raw:" + rawValue);

    pin_value = analogRead(VBAT_PIN);
    float voltage_level = (pin_value / static_cast<float>(4096)) * 2 * 1.115 * 3.3; // calculate voltage level
   //battery_level = (voltage_level / BATTV_MAX * 100);
    //Serial.println((String)"Raw:" + pin_value + " Voltage:" + voltage_level + "V Percent: " + (battery_level)+"%");


    if (pin_value >= 2400) battery_level = 100;
    else if (pin_value >= 2100)    // 85-100% range
        battery_level = map(pin_value, 2100, 2400, 85, 100);
    else if (pin_value >= 1800)    // 10-85% range
        battery_level = map(pin_value, 1800, 2100, 10, 85);
    else if (pin_value >= 1700)    // 0-10% range
        battery_level = map(pin_value, 1700, 1800, 0, 10);
    else battery_level = 0;

    Serial.println((String)"Raw:" + pin_value + " Voltage:" + voltage_level + "V Percent: " + (battery_level)+"%");
}

void startSleep() {
    //Increment boot number and print it every reboot
    ++bootCount;
    Serial.println("Boot number: " + String(bootCount));

    esp_sleep_enable_timer_wakeup(TIME_TO_SLEEP * uS_TO_S_FACTOR);
    Serial.println("Setup ESP32 to sleep for every " + String(TIME_TO_SLEEP) + " Seconds");
    Serial.println("Going to sleep now");
    delay(1000);
    Serial.flush();
    esp_deep_sleep_start();

}

void initLoRaSender() {
    Serial.println("LoRa Sender");


    Serial.print(F("[SX1276] Initializing ... "));
    int state = radio.beginFSK();
    radio.setFrequency(868.0);

    if (state == RADIOLIB_ERR_NONE) {
        Serial.println(F("success!"));
    }
    else {
        Serial.print(F("failed, code "));
        Serial.println(state);
        while (true);
    }


}

void loop() {
    manageTemperature();

}

void sendLoRaMessage() {
    Serial.print("Temp value: ");
    Serial.println(temperatureMessage);
    String result =  sha1(temperatureMessage); // Creating the hash value for our string

    // send packet
    int state = radio.transmit(result);
    Serial.println("Package sent");

    if (state == RADIOLIB_ERR_NONE) {
        // the packet was successfully transmitted
        Serial.println(F(" success!"));

        // print measured data rate
        Serial.print(F("[SX1276] Datarate:\t"));
        Serial.print(radio.getDataRate());
        Serial.println(F(" bps"));

    }
    else if (state == RADIOLIB_ERR_PACKET_TOO_LONG) {
        // the supplied packet was longer than 256 bytes
        Serial.println(F("too long!"));

    }
    else if (state == RADIOLIB_ERR_TX_TIMEOUT) {
        // timeout occurred while transmitting packet
        Serial.println(F("timeout!"));

    }
    else {
        // some other error occurred
        Serial.print(F("failed, code "));
        Serial.println(state);

    }
}

void initSerial() {
    Serial.begin(115200);//9600 before
    Serial.setTimeout(2000);

    // Wait for serial to initialize.
    while (!Serial) {}
}


void initDHT22() {
    dht.begin();

    Serial.println("Device Started");
    Serial.println("-------------------------------------");
    Serial.println("Running DHT!");
    Serial.println("-------------------------------------");
}

void manageTemperature() {

    memset(temperatureMessage, 0, sizeof(temperatureMessage));

    DynamicJsonDocument doc(1024);

    while  (tempAvailable != 1)
    {
        // Report every 2 seconds.
        if (timeSinceLastRead > 2000) {
            // Reading temperature or humidity takes about 250 milliseconds!
            // Sensor readings may also be up to 2 seconds 'old' (its a very slow sensor)
            float h = dht.readHumidity();
            // Read temperature as Celsius (the default)
            float t = dht.readTemperature();

            // Check if any reads failed and exit early (to try again).
            if (isnan(h) || isnan(t)) {
                Serial.println("Failed to read from DHT sensor!");
                timeSinceLastRead = 0;
            }
            // Compute heat index in Celsius (isFahreheit = false)
            float hic = dht.computeHeatIndex(t, h, false);

            Serial.print("Humidity: ");
            Serial.print(h);
            Serial.print(" %\t");
            Serial.print("Temperature: ");
            Serial.print(t);
            Serial.print(" *C ");
            Serial.print("Heat index: ");
            Serial.print(hic);
            Serial.println(" *C ");

            timeSinceLastRead = 0;

            doc["sensor"] = "temperature";
            doc["sensor_model"] = "DHT22";
            doc["location"] = "cantina";
            doc["humidy"] = String(h);
            doc["temperature"] = String(t);
            doc["heat_index"] = String(hic);
            doc["battery_level"] = String(battery_level);
            doc["pin_value"] = String(pin_value);

            serializeJson(doc, temperatureMessage);
            Serial.println(temperatureMessage);
        
            sendLoRaMessage();
            delay(1000);

            tempAvailable = 1;
            
        }
        delay(1000);
        timeSinceLastRead += 100;
    }
    tempAvailable = 0;
    
}
```

**Sender output**
```
LoRa Sender
08:00:48:171 -> [SX1276] Initializing ... success!
Temp value: {"sensor":"temperature","sensor_model":"DHT22","location":"cantina","humidy":"60.90","temperature":"26.20","heat_index":"27.18","battery_level":"70","pin_value":"2041"}
Package sent
 success!
[SX1276] Datarate:	53333332.00 bps
```

### **Receiver**
```c++
#include <Arduino.h>
#include <SPI.h>
#include <Wire.h>
#include <RadioLib.h>


/****
This section is used for the LoRa setup
*****/
#define LoRa_MOSI 10
#define LoRa_MISO 11
#define LoRa_SCK 9
#define LoRa_nss 8
#define LoRa_dio1 14
#define LoRa_nrst 12
#define LoRa_busy 13
SX1262 radio = new Module(LoRa_nss, LoRa_dio1, LoRa_nrst, LoRa_busy);


//global var
unsigned long previousMillis = 0;
unsigned long lastRestartMillis = 0;
unsigned long interval = 60000;


void setup() {
    Serial.begin(115200);
    while (!Serial);

      // initialize SX1262 with default settings
  Serial.print(F("[SX1262] Initializing ... "));
  int state = radio.begin();
  radio.setFrequency(868.0);
  
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
  Serial.print(F("[SX1262] Waiting for incoming transmission ... "));
}

void loop() {
  

  // you can receive data as an Arduino String
  String str;
  int state = radio.receive(str);

  // you can also receive data as byte array
  /*
    byte byteArr[8];
    int state = radio.receive(byteArr, 8);
  */

  if (state == RADIOLIB_ERR_NONE) {
    // packet was successfully received
    Serial.println(F("success!"));

    // print the data of the packet
    Serial.print(F("[SX1262] Data:\t\t"));
    Serial.println(str);

    // print the RSSI (Received Signal Strength Indicator)
    // of the last received packet
    Serial.print(F("[SX1262] RSSI:\t\t"));
    Serial.print(radio.getRSSI());
    Serial.println(F(" dBm"));

    // print the SNR (Signal-to-Noise Ratio)
    // of the last received packet
    Serial.print(F("[SX1262] SNR:\t\t"));
    Serial.print(radio.getSNR());
    Serial.println(F(" dB"));

    // print frequency error
    Serial.print(F("[SX1262] Frequency error:\t"));
    Serial.print(radio.getFrequencyError());
    Serial.println(F(" Hz"));

  } else if (state == RADIOLIB_ERR_RX_TIMEOUT) {
    // timeout occurred while waiting for a packet
    //Serial.println(F("timeout!"));

  } else if (state == RADIOLIB_ERR_CRC_MISMATCH) {
    // packet was received, but is malformed
    Serial.println(F("CRC error!"));

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);

  }
}
```

**Receiver Output**
```
[SX1262] Initializing ... success!
[SX1262] Waiting for incoming transmission ... 
```

**Comment by jgromes** [comment]
Your SX1276 is transmitting FSK and your SX126x is receiving LoRa, that will not work. You have to choose one of those.

Also, the measured data rate is reported as about 50 Mbps - not a realistic number, which would seem to suggest the message you're trying to send could be empty.

**Comment by lucadjc** [comment]
Ciao,
SX1276 changing to 

int state = radio.begin();
radio.setFrequency(868.0);

the result is
09:08:09:587 -> 
{"sensor":"temperature","sensor_model":"DHT22","location":"cantina","humidy":"65.40","temperature":"26.00","heat_index":"27.13","battery_level":"70","pin_value":"2041"}
09:08:10:161 -> timeout!

Tested even with a simple message "Ciao" and the result is timeout too.

**Comment by jgromes** [comment]
If your transmitter is timing out, it suggest a wiring issue. This line is rather suspicious:

`SX1276 radio = new Module(ss, dio1, rst, dio0);` DIO0 should come first.

Thie doesn't seem t obe an issue in the library itself, hence I'm converting this thread to a discussion instead.

**Reply by lucadjc** [reply]
ciao, many thanks for your support.
i changed the order, now I've

// - Pin configs - ESP32 -
#define ss 18   
#define rst  14
#define dio0 26 
#define dio1 35 
SX1276 radio = new Module(ss, dio0, rst,  dio1);

this is the **[SX1276] output**, it seems better :)
"sensor":"temperature","sensor_model":"DHT22","location":"cantina","humidy":"50.60","temperature":"25.60","heat_index":"25.54","battery_level":"70","pin_value":"2042"}
Temp value: {"sensor":"temperature","sensor_model":"DHT22","location":"cantina","humidy":"50.60","temperature":"25.60","heat_index":"25.54","battery_level":"70","pin_value":"2042"}
 success!
[SX1276] Datarate:	865.80 bps

on SX1262 I read this discussion https://github.com/jgromes/RadioLib/discussions/598 and i made the following change

SPIClass spi(FSPI); //here I used FSPI because VSPI has not be found in compiling phase
SPISettings spiSettings(2000000, MSBFIRST, SPI_MODE0);
SX1262 radio = new Module(8,14,12,13, spi, spiSettings);
...
spi.begin(9, 11, 10, 8); //SCK, MISO, MOSI, SS
Serial.println(F("[SX1262] Initializing ... "));
int state = radio.begin();
radio.setFrequency(868.0);

this is the **[SX1276] output**, 
[SX1262] Data:		0fa60363811526877ba1948ceb043de367dd4e64
[SX1262] RSSI:		-79.00 dBm
[SX1262] SNR:		10.75 dB
[SX1262] Frequency error:	9580.94 Hz

Now I receive the message but a lot are lost. the sender is messaging around 1 time per minutes but the receiver is "taking" around 10%.

suggestions?

**Comment by lucadjc** [comment, marked as answer]
Hi,
I've resolved the issue changing the SX1262 code as following

#include <Arduino.h>
#include <SPI.h>
#include <Wire.h>
#include <RadioLib.h>


/****
This section is used for the LoRa setup
*****/
#define LoRa_MOSI 10
#define LoRa_MISO 11
#define LoRa_SCK 9
#define LoRa_nss 8
#define LoRa_dio1 14
#define LoRa_nrst 12
#define LoRa_busy 13
//SX1262 radio = new Module(LoRa_nss, LoRa_dio1, LoRa_nrst, LoRa_busy);


//#define VSPI FSPI

SPIClass spi(FSPI);
SPISettings spiSettings(100000, MSBFIRST, SPI_MODE0);
SX1262 radio = new Module(LoRa_nss,LoRa_dio1,LoRa_nrst,LoRa_busy, spi, spiSettings);

//global var
unsigned long previousMillis = 0;
unsigned long lastRestartMillis = 0;
unsigned long interval = 60000;
const char* version = "1.6";
char  ROWS[6][24];

void(*resetFunc) (void) = 0;


void setup() {
    Serial.begin(115200);
    while (!Serial);

      // initialize SX1262 with default settings
  spi.begin(LoRa_SCK, LoRa_MISO, LoRa_MOSI, LoRa_nss); //SCK, MISO, MOSI, SS
  Serial.println(F("[SX1262] Initializing ... "));
  int state = radio.begin();
  radio.setFrequency(868.0);
  
  Serial.print(F("Output code "));
    Serial.println(state);

  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
  Serial.print(F("[SX1262] Waiting for incoming transmission ... "));
  // set the function that will be called
  // when new packet is received
  radio.setPacketReceivedAction(setFlag);

    // start listening for LoRa packets
  Serial.print(F("[SX1262] Starting to listen ... "));
  state = radio.startReceive();
  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
}

// flag to indicate that a packet was received
volatile bool receivedFlag = false;

// this function is called when a complete packet
// is received by the module
// IMPORTANT: this function MUST be 'void' type
//            and MUST NOT have any arguments!
#if defined(ESP8266) || defined(ESP32)
  ICACHE_RAM_ATTR
#endif
void setFlag(void) {
  // we got a packet, set the flag
  receivedFlag = true;
}

void loop() {
  // check if the flag is set
  if(receivedFlag) {
    // reset flag
    receivedFlag = false;

    // you can read received data as an Arduino String
    String str;
    int state = radio.readData(str);

    // you can also read received data as byte array
    /*
      byte byteArr[8];
      int state = radio.readData(byteArr, 8);
    */

    if (state == RADIOLIB_ERR_NONE) {
      // packet was successfully received
      Serial.println(F("[SX1262] Received packet!"));

      // print data of the packet
      Serial.print(F("[SX1262] Data:\t\t"));
      Serial.println(str);

      // print RSSI (Received Signal Strength Indicator)
      Serial.print(F("[SX1262] RSSI:\t\t"));
      Serial.print(radio.getRSSI());
      Serial.println(F(" dBm"));

      // print SNR (Signal-to-Noise Ratio)
      Serial.print(F("[SX1262] SNR:\t\t"));
      Serial.print(radio.getSNR());
      Serial.println(F(" dB"));

      // print frequency error
      Serial.print(F("[SX1262] Frequency error:\t"));
      Serial.print(radio.getFrequencyError());
      Serial.println(F(" Hz"));

    } else if (state == RADIOLIB_ERR_CRC_MISMATCH) {
      // packet was received, but is malformed
      Serial.println(F("CRC error!"));

    } else {
      // some other error occurred
      Serial.print(F("failed, code "));
      Serial.println(state);

    }
  }
}
